documentation/public/user-guide/concepts.md

```diff
@@ -1,11 +1,52 @@
+This is a departure from the classic approach of "black-box" compilers, which are handed the input and only their output can be observed.
 
 ## Concrete Syntax Tree (CST)
 
-At the time of writing, Slang is capable of parsing the source code into a Concrete Syntax Tree (CST; also sometimes called "full-fidelity"), which is a tree structure of the program that also includes things like punctuation or whitespace.
+At the time of writing, Slang is capable of parsing the source code into a Concrete Syntax Tree (CST; also sometimes called "full-fidelity"),
+which is a tree structure of the program that also includes things like punctuation or whitespace.
+
+This is done by using the (standard) approach of lexical analysis followed by syntax analysis.
+The source text as a sequence of characters is recognized into a sequence of tokens (lexical analysis), which then in turn is _parsed_ into the CST.
+
+The resulting CST is a regular tree data structure that you can visit.
+The tree nodes are represented by the `Node` structure, which can be one of two kinds:
+
+-   `RuleNode` (aka _non-terminals_) represent sub-trees, containing a a vector of other `Node` children.
+-   `TokenNode` (aka _terminals_) are leaves and represent a lexical token (i.e. an identifier, keyword, punctuation) in the source.
+
+## Parser API
+
+A parser API is created by using a language version to create a `Language` object.
+You can then use it to parse different inputs belonging to that version.
+
+Each `parse()` operation accepts the input source code, and a `RuleKind` variant.
+This allows callers to parse entire source files (`RuleKind::SourceUnit`), individual contracts (`RuleKind::ContractDefinition`),
+methods (`RuleKind::FunctionDefinition`), or any other syntax nodes.
+
+The resulting `ParseOutput` object will contain syntax errors (if any), and the parse tree (CST) corresponding to the input source code.
+
+## Cursor API
+
+For many code analysis tasks, it is useful to traverse the parse tree and visit each node.
+The `Cursor` object allows callers to traverse the parse tree in an efficient pre-order manner.
+
+It provides several `goTo*()` navigation functions, each returning `true` if the cursor was successfully moved, and `false` otherwise.
+There are three main ways to do it:
+
+-   According to the DFS order, i.e. `goToNext()` and `goToPrevious()`,
+-   According to the relationship between the current node and the next node, i.e. `goToParent()`, `goToFirstChild()`, `goToNextNonDescendent()`
+-   According to the kind of the next node, i.e. `goToNextTokenWithKind(kind)`, `goToNextRuleWithKind(kind)`
+
+As such, the cursor is stateful and keeps track of the path it has taken through the CST.
+It starts at the root it was created at and is completed when it reaches its root when navigating forward.
```

Comment:
The "XYZ API" sections here feel very out of place for me - I'd consider "concepts" to be a higher-level overview of how data flows, while both of these feel like they should be in the first/introductory paragraph in guides dedicated specifically to them, with how code/backtick-heavy and precise/concrete they are.

I'd like to propose adding a H3 to the CST (or a H2 immediately after CST) with a basic outline: effectively removing the "It provides several (...)" paragraph and rewording it for simplicity/high-level overview.

As for the Parser API, maybe we could have a dedicated H2 "Exposing the APIs/API Layers" (much like https://github.com/dotnet/roslyn/blob/main/docs/wiki/Roslyn-Overview.md#api-layers) at the end, that includes it?

This way, we would have high-level concepts (CST/CST Cursor/AST) followed by interfaces/usage that exposes them, like the Parser API.

Reply:
I think the Roslyn example makes sense if we have 20 different APIs across multiple levels of layering to describe :) Our document here is much simpler.
But I do see your point about keeping this focused on concepts rather than APIs. Let me restructure this a bit.

Reply:
I think the Roslyn example makes sense once we start to have tens of APIs across multiple layers of abstractions. But for this, we have a few mere components that have a clear connection between them, and users would naturally want to learn about them in order, as you cannot use one without the previous.
But I do see your point about focusing on components rather than APIs. let me reword the headers a bit.